File: cs101-Algorithms_and_ADTs/asg4-C-BFS_and_Shortest_Paths_in_Graphs/List.c
/*******************************************************************************
 *  $Id: List.c,v 1.2 2010-11-14 19:30:00-08 dmfrank - $
 *	Derek Frank, dmfrank@ucsc.edu
 *	
 *	NAME
 *	  List.c
 *  
 *	DESCRIPTION
 *	  Implementation file for the List ADT.  A linked list for integers with
 *	markers to keep track of the front and rear of the double ended queue.
 *	Includes a current-position marker that can move between previous and next
 *	positions.
 ******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include "List.h"

/************** Private Structs: not exported *********************************/

typedef struct Node {
	int data;
	struct Node* next;
	struct Node* prev;
} Node;

typedef Node* NodeRef;

typedef struct List {
	NodeRef front;
	NodeRef back;
	NodeRef curr;
	int length;
} List;


/************** Constructors-Destructors **************************************/

/*
 *  newNode
 *  Returns pointer to new Node struct. Initializes next field to NULL, and sets
 *  data field to input parameter node_data.
 *	Private.
 */
NodeRef newNode(int node_data) {
	NodeRef N = malloc(sizeof(Node));
	N->data = node_data;
	N->next = NULL;
	N->prev = NULL;
	return (N);
}

/*
 *  freeNode
 *  Frees heap memory pointed to by pN.
 *	Private.
 */
void freeNode(NodeRef* pN) {
	if ( pN != NULL && *pN != NULL ) {
		free(*pN);
		*pN = NULL;
	}
}


/*
 *  newList()
 *  Returns ListRef pointing to new List Struct which represents an empty List.
 *  Initializes front, current, and back fields to NULL, sets length field to 0.
 *	Exported.
 */
ListRef newList(void) {
	ListRef L;
	L = malloc(sizeof(List));
	L->front = L->back = L->curr = NULL;
	L->length = 0;
	return (L);
}

/*
 *  freeList()
 *  Frees all heap memory associated with the ListRef *pL, including all memory
 *  in existing Nodes.  Sets *pL to NULL.
 *	Exported.
 *	Post: isEmpty(pL)
 */
void freeList(ListRef* pL) {
	if ( pL == NULL || *pL == NULL ) {
		return;
	}
	while ( !isEmpty(*pL) ) {
		deleteFront(*pL);
	}
	free(*pL);
	*pL = NULL;
}



/***************** Access functions *******************************************/

/*
 *  isEmpty()
 *  Returns True if L is empty, otherwise returns false.
 */
int isEmpty(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling isEmpty() on NULL ListRef\n");
		exit(1);
	}
	return (L->length == 0);
}

/*
 *  offEnd()
 *  Returns True if curr is NULL in L, otherwise returns false.
 */
int offEnd(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling offEnd() on NULL ListRef\n");
		exit(1);
	}
	return (L->curr == NULL);
}

/*
 *  atFront()
 *  Returns True if curr is pointing at the front of L, otherwise returns false.
 *	Pre: !isEmpty(L); !offEnd(L).
 */
int atFront(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling atFront() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling atFront() on an empty List\n");
		exit(1);
	}
	if ( offEnd(L) ) {
		printf("List Error: calling atFront() on NULL pointer\n");
		exit(1);
	}
	return (L->curr->prev == NULL);
}

/*
 *  atBack()
 *  Returns True if curr is pointing at the back of L, otherwise returns false.
 *	Pre: !isEmpty(L); !offEnd(L).
 */
int atBack(ListRef L) {
	if ( L == NULL ){
		printf("List Error: calling atBack() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling atBack() on an empty List\n");
		exit(1);
	}
	if ( offEnd(L) ) {
		printf("List Error: calling atBack() on NULL pointer\n");
		exit(1);
	}
	return (L->curr->next == NULL);
}

/*
 *  getFront()
 *  Returns the value at the front of L.
 *  Pre: !isEmpty(L)
 */
int getFront(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling getFront() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling getFront() on an empty List\n");
		exit(1);
	}
	return (L->front->data);
}

/*
 *  getBack()
 *  Returns the value at the back of L.
 *  Pre: !isEmpty(L)
 */
int getBack(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling getBack() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling getBack() on an empty List\n");
		exit(1);
	}
	return (L->back->data);
}

/*
 *  getCurrent()
 *  Returns the value pointed at by curr in L.
 *  Pre: !isEmpty(L); !offEnd(L).
 */
int getCurrent(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling getCurrent() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling getCurrent() on an empty List\n");
		exit(1);
	}
	if ( offEnd(L) ) {
		printf("List Error: calling getCurrent() on NULL pointer\n");
		exit(1);
	}
	return (L->curr->data);
}

/*
 *  getLength()
 *  Returns the length of L.
 */
int getLength(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling getLength() on NULL ListRef\n");
		exit(1);
	}
	return (L->length);
}

/*
 *  equals()
 *  Returns true if A is identical to B, false otherwise.  Ignores current
 *	marker position.
 *  Pre: !isEmpty(A,B)
 */
int equals(ListRef A, ListRef B) {
	int flag = 1;
	NodeRef N = NULL;
	NodeRef M = NULL;
	
	if ( A==NULL || B==NULL ) {
		printf("List Error: calling equals() on NULL ListRef\n");
		exit(1);
	}
	N = A->front;
	M = B->front;
	if ( getLength(A) == getLength(B) ) {
		while ( flag && N != NULL ) {
			flag = (N->data == M->data);
			N = N->next;
			M = M->next;
		}
		return flag;
	}
	return (0);
}


/**************** Manipulation procedures *************************************/

/*
 *  makeEmpty()
 *  Sets L to the empty state.
 *	Post: isEmpty(L).
 */
void makeEmpty(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling makeEmpty() on NULL ListRef\n");
		exit(1);
	}
	if ( L->front == NULL) {
		return;
	}
	while( !isEmpty(L) ) {
		deleteFront(L);
	}
}

/*
 *  moveFront()
 *  Sets current marker to front.
 *	Pre: !isEmpty(L).
 */
void moveFront(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling moveFront() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling moveFront() on an empty List\n");
		exit(1);
	}
	L->curr = L->front;
}

/*
 *  moveLast()
 *  Sets current marker to back.
 *	Pre: !isEmpty(L).
 */
void moveLast(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling moveLast() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling moveBack() on an empty List\n");
		exit(1);
	}
	L->curr = L->back;
}

/*
 *  movePrev()
 *  Moves current marker one step toward front.
 *	Pre: !isEmpty(L); !offEnd(L).
 */
void movePrev(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling movePrev() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling movePrev() on an empty List\n");
		exit(1);
	}
	if ( offEnd(L) ) {
		printf("List Error: calling movePrev() on a NULL pointer\n");
		exit(1);
	}
	L->curr = L->curr->prev;
}

/*
 *  moveNext()
 *  Moves current marker one step toward last element.
 *	Pre: !isEmpty(L); !offEnd(L).
 */
void moveNext(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling moveNext() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling moveNext() on an empty List\n");
		exit(1);
	}
	if ( offEnd(L) ) {
		printf("List Error: calling moveNext() on a NULL pointer\n");
		exit(1);
	}
	L->curr = L->curr->next;
}

/*
 *  insertBeforeFront()
 *  Places new data element at the front of L.
 *  Post: !isEmpty(L)
 */
void insertBeforeFront(ListRef L, int data) {
	if ( L == NULL ) {
		printf("List Error: calling insertBeforeFront() on NULL ListRef\n");
		exit(1);
	}
	
	NodeRef N = newNode(data);
	if ( isEmpty(L) ) {
		L->front = L->back = N;
	}
	else {
		N->next = L->front;
		L->front->prev = N;
		L->front = N;
	}
	++L->length;
}

/*
 *  insertAfterBack()
 *  Places new data element at the back of L.
 *  Post: !isEmpty(L)
 */
void insertAfterBack(ListRef L, int data) {
	if ( L == NULL ) {
		printf("List Error: calling insertAfterBack() on NULL ListRef\n");
		exit(1);
	}
	
	NodeRef N = newNode(data);
	if ( isEmpty(L) ) {
		L->front = L->back = N;
	}
	else {
		L->back->next = N;
		N->prev = L->back;
		L->back = N;
	}
	++L->length;
}

/*
 *  insertBeforeCurrent()
 *  Places new data element before curr in L.
 *	Pre: !isEmpty(L); !offEnd(L)
 */
void insertBeforeCurrent(ListRef L, int data) {
	if ( L == NULL ) {
		printf("List Error: calling insertBeforeCurrent() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling insertBeforeCurrent() on an empty List\n");
		exit(1);
	}
	if ( offEnd(L) ) {
		printf("List Error: calling insertBeforeCurrent() on a NULL pointer\n");
		exit(1);
	}
	
	NodeRef N = newNode(data);
	if ( L->curr->prev == NULL ) {
		N->next = L->front;
		L->front->prev = N;
		L->front = N;
	}else {
		N->prev = L->curr->prev;
		N->next = L->curr;
		L->curr->prev->next = N;
		L->curr->prev = N;
	}
	++L->length;
}

/*
 *  insertAfterCurrent()
 *  Places new data element after curr in L.
 *  Pre: !isEmpty(L); !offEnd()
 */
void insertAfterCurrent(ListRef L, int data) {
	if ( L == NULL ) {
		printf("List Error: calling insertAfterCurrent() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling insertAfterCurrent() on an empty List\n");
		exit(1);
	}
	if ( offEnd(L) ) {
		printf("List Error: calling insertAfterCurrent() on a NULL pointer\n");
		exit(1);
	}
	
	NodeRef N = newNode(data);
	if ( L->curr->next == NULL ) {
		L->back->next = N;
		N->prev = L->back;
		L->back = N;
	}else {
		N->prev = L->curr;
		N->next = L->curr->next;
		L->curr->next->prev = N;
		L->curr->next = N;
	}
	++L->length;
}

/*
 *  deleteFront()
 *  Deletes element at front of L.
 *  Pre: !isEmpty(L)
 */
void deleteFront(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling deleteFront() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling deleteFront() on an empty List\n");
		exit(1);
	}
	
	NodeRef N = NULL;
	N = L->front;
	if ( getLength(L) > 1 ) {
		L->front = L->front->next;
		L->front->prev = NULL;
	}else {
		L->front = L->back = L->curr = NULL;
	}
	--L->length;
	freeNode(&N);
}

/*
 *  deleteBack()
 *  Deletes element at back of L.
 *  Pre: !isEmpty(L)
 */
void deleteBack(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling deleteBack() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling deleteBack() on an empty List\n");
		exit(1);
	}
	
	NodeRef N = NULL;
	N = L->back;
	if ( getLength(L) > 1 ) {
		L->back = L->back->prev;
		L->back->next = NULL;
	}else {
		L->front = L->back = L->curr = NULL;
	}
	--L->length;
	freeNode(&N);
}

/*
 *  deleteCurrent()
 *  Deletes element at curr in L.
 *  Pre: !isEmpty(L); !offEnd(L)
 */
void deleteCurrent(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling deleteCurrent() on NULL ListRef\n");
		exit(1);
	}
	if ( isEmpty(L) ) {
		printf("List Error: calling deleteCurrent() on an empty List\n");
		exit(1);
	}
	if ( offEnd(L) ) {
		printf("List Error: calling deleteCurrent() on a NULL pointer\n");
		exit(1);
	}
	
	NodeRef N = NULL;
	N = L->curr;
	if ( getLength(L) > 1 ) {
		if ( L->curr->prev == NULL ) {
			L->front = L->curr->next;
			L->curr->next = NULL;
		}else if ( L->curr->next == NULL ) {
			L->back = L->curr->prev;
			L->curr->prev = NULL;
		}else {
			L->curr->prev->next = L->curr->next;
			L->curr->next->prev = L->curr->prev;
			L->curr->prev = L->curr->next = NULL;
		}
		L->curr = NULL;
	}else {
		L->front = L->back = L->curr = NULL;
	}
	--L->length;
	freeNode(&N);
}


/*************** Other Functions **********************************************/
/*
 *  printList()
 *  Prints data elements in L on a single line to specified output file.
 */
void printList(FILE* out, ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling printList() on NULL ListRef\n");
		exit(1);
	}

	NodeRef N = NULL;
	for ( N = L->front; N != NULL; N = N->next ) {
		fprintf(out, "%d ", N->data);
	}
	fprintf(out, "\n");
}

/*
 *  copyList()
 *  Returns a ListRef to a new list equivalent to L.
 */
ListRef copyList(ListRef L) {
	if ( L == NULL ) {
		printf("List Error: calling copyList() on NULL ListRef\n");
		exit(1);
	}
	
	ListRef copy = newList();
	NodeRef N = L->front;
	while ( N != NULL) {
		insertAfterBack(copy, N->data);
		N = N->next;
	}
	return (copy);
}

/*
 *	catList()
 *	Returns a new List that is the concatenation of List A followed by List B.
 *	The current marker in the new list is undefined regardless of the states of
 *	the current markers in the two lists.  The states of the two Lists are
 *	unchanged.
 */
ListRef catList(ListRef A, ListRef B) {
	if ( A == NULL || B == NULL) {
		printf("List Error: calling catList() on NULL ListRef\n");
		exit(1);
	}

	ListRef mainList = copyList(A);
    ListRef catList = copyList(B);
    mainList->back->next = catList->front;
    catList->front->prev = mainList->back;
    mainList->back = catList->back;
    mainList->length += catList->length;
	catList = NULL;
    return (mainList);
}

File: cs101-Algorithms_and_ADTs/asg3-Java-Sparse_Matrix_ADT/List.java
/* $Id: List.java,v 1.2 2010-10-31 07:26:38-07 dmfrank - $
 * Derek Frank, dmfrank@ucsc.edu
 * 
 * NAME
 *   List -- a double ended Object queue with a current-position 
 * marker.
 * 
 * DESCRIPTION
 *   A linked list for Objects with markers to keep track of the
 * front and rear of the double ended queue.  Includes a current-
 * position marker that can move to previous and next positions.
 */

class List {
  
  // Inner Classes //////////////////////////////////////////////
  
  /* class Node
   * Private inner class
   */
  private class Node{
    
    // Fields //
    Object data;
    Node next;
    Node prev;
    
    // Constructor //
    
    /* Node()
     * Create new node.
     */
    Node(Object data) {
      this.data = data;
      next = prev = null;
    }
    
    // Access Functions //
    
    /* equals()
     * Overrides Object's equals() method.
     * Pre: (obj instanceof Node)
     */
    public boolean equals(Object obj) {
      if ( !(obj instanceof Node) ) return false;
      if (obj == this) return true;
      Node N = (Node)obj;
      
      Node M = this;
      return N.data.equals(M.data);
    }
    
    // Other Functions //
    
    /* toString()
     * Overides Object's toString method.
     */
    public String toString() {
      String str = "";
      str += this.data.toString();
      return str;
    }
    
  }
  
  // Fields /////////////////////////////////////////////////////
  private Node front;
  private Node back;
  private Node curr;
  private int length;
  
  // Constructors ///////////////////////////////////////////////
  
  /* List()
   * Create new empty List.
   */
  List() {
    front = back = curr = null;
    length = 0;
  }
  
  // Access functions ///////////////////////////////////////////
  
  /* isEmpty()
   * Returns true is this List is empty, false otherwise.
   */
  boolean isEmpty() {
    return (length == 0);
  }
  
  /* offEnd()
   * Returns true if current is undefined.
   */
  boolean offEnd() {
    return (this.curr == null);
  }
  
  /* atFirst()
   * Returns true if first element is current.
   * Precondition: !isEmpty().
   */
  boolean atFirst() { 
    if ( this.isEmpty() || this.offEnd() ) return false;
    return (this.curr.prev == null);
  }
  
  /* atLast()
   * Returns true if last element is current.
   * Pre: !isEmpty().
   */
  boolean atLast() {
    if ( this.isEmpty() || this.offEnd() ) return false;
    return (curr.next == null);
  }
  
  /* getFirst()
   * Returns first element.
   * Pre: !isEmpty().
   */
  Object getFirst() {
    if( this.isEmpty() ) {
      throw new RuntimeException("List Error: getFirst() called on empty List");
    }
    return front.data;
  }
  
  /* getLast()
   * Returns last element.
   * Pre: !isEmpty().
   */
  Object getLast() {
    if ( this.isEmpty() ) {
      throw new RuntimeException("List Error: getLast() called on empty List");
    }
    return back.data;
  }
  
  /* getCurrent()
   * Returns current element.
   * Pre: !isEmpty(), !offEnd().
   */
  Object getCurrent() {
    if ( this.isEmpty() ) {
      throw new RuntimeException("List Error: getCurrent() called on empty List");
    }else if ( this.offEnd() ) {
      throw new RuntimeException("List Error: getCurrent() called on Null Pointer");
    }
    return curr.data;
  }
  
  /* getLength()
   * Returns length of this List.
   */
  int getLength() { return length; }
  
  /* equals()
   * Overrides Object's equals() method.
   * Pre: (obj instanceof List)
   */
  public boolean equals(Object obj) {
    if ( !(obj instanceof List) ) return false;
    if (obj == this) return true;
    List L = (List)obj;
    
    boolean flag  = true;
    Node N = this.front;
    Node M = L.front;
    
    if ( this.length == L.length ) {
      while ( flag && N != null ) {
        flag = ( N.equals(M) );
        N = N.next;
        M = M.next;
      }
      return flag;
    }else {
      return false;
    }
  }
  
  // Manipulation procedures ////////////////////////////////////
  
  /* makeEmpty()
   * Sets this List to the empty state.
   * Post: isEmpty().
   */
  void makeEmpty() {
    this.back = null;
    while ( !this.isEmpty() ) {
      this.curr = this.front;
      this.curr = this.curr.next;
      if ( this.curr != null ) this.curr.prev = null;
      this.front.next = null;
      --length;
    }
    this.front = null;
  }
  
  /* moveFirst()
   * Sets current marker to first element.
   * Pre: !isEmpty(); Post: !offEnd().
   */
  void moveFirst() {
    if ( this.isEmpty() ) {
      throw new RuntimeException("List Error: moveFirst() called on empty List");
    }else if ( this.offEnd() ) { this.curr = this.front; }
    else {
      while ( this.curr.prev != null ) {
        this.curr = this.curr.prev;
      }
    }
  }
  
  /* moveLast()
   * Sets current marker to last element
   * Pre: !isEmpty(); Post: !offEnd().
   */
  void moveLast() {
    if ( this.isEmpty() ) {
      throw new RuntimeException("List Error: moveLast() called on empty List");
    }else if ( this.offEnd() ) { this.curr = this.back; }
    else {
      while ( this.curr.next != null ) {
        this.curr = this.curr.next;
      }
    }
  }
  
  /*movePrev()
   * Moves current marker one step toward first element.
   * Pre: !isEmpty(), !offEnd().
   */
  void movePrev() {
    if ( this.isEmpty() ) {
      throw new RuntimeException("List Error: movePrev() called on empty List");
    }else if ( this.offEnd() ) {
      throw new RuntimeException("List Error: movePrev() called on Null Pointer");
    }else { this.curr = this.curr.prev; }
  }
  
  /* moveNext()
   * Moves current marker one step toward last element.
   * Pre: !isEmpty(), !offEnd().
   */
  void moveNext() {
    if ( this.isEmpty() ) {
      throw new RuntimeException("List Error: moveNext() called on empty List");
    }else if ( this.offEnd() ) {
      throw new RuntimeException("List Error: moveNext() called on Null Pointer");
    }else { this.curr = this.curr.next; }
  }
  
  /* insertBeforeFirst()
   * Inserts new element before first element.
   * Post: !isEmpty().
   */
  void insertBeforeFirst(Object data) {
    Node N = new Node(data);
    if ( this.isEmpty() ) { front = back = N; }
    else { N.next = front; front.prev = N; front = N; }
    ++length;
  }
  
  /* insertAfterLast()
   * Inserts new element after last element.
   * Post: !isEmpty().
   */
  void insertAfterLast(Object data) {
    Node N = new Node(data);
    if ( this.isEmpty() ) { front = back = N; }
    else { N.prev = back; back.next = N; back = N; }
    ++length;
  }
  
  /* insertBeforeCurrent()
   * Inserts new element before current element.
   * Pre: !isEmpty(), !offEnd().
   */
  void insertBeforeCurrent(Object data){
    if ( this.isEmpty() ) {
      throw new RuntimeException("ListError: insertBeforeCurrent() called on empty List");
    }else if ( this.offEnd() ) {
      throw new RuntimeException("ListError: insertBeforeCurrent() called on a Null Pointer");
    }
    Node N = new Node(data);
    if ( curr.prev == null) {
      N.next = front;
      front.prev = N;
      front = N;
    }else {
      N.prev = curr.prev;
      N.next = curr;
      curr.prev.next = N;
      curr.prev = N;
    }
    ++length;
  }
  
  /* insertAfterCurrent()
   * Inserts new element after current element.
   * Pre: !isEmpty(), !offEnd().
   */
  void insertAfterCurrent(Object data) {
    if ( this.isEmpty() ) {
      throw new RuntimeException("ListError: insertAfterCurrent() called on empty List");
    }else if ( this.offEnd() ) {
      throw new RuntimeException("ListError: insertAfterCurrent() called on a Null Pointer");
    }
    Node N = new Node(data);
    if ( curr.next == null) {
      back.next = N;
      back = N;
      back.prev = curr;
    }else {
      N.prev = curr;
      N.next = curr.next;
      curr.next.prev = N;
      curr.next = N;
    }
    ++length;
  }
  
  /* deleteFirst()
   * Deletes first element.
   * Pre: !isEmpty().
   */
  void deleteFirst() {
    if( this.isEmpty() ){
      throw new RuntimeException("List Error: deleteFirst() called on empty List");
    }
    if ( this.length > 1 ) {
      front = front.next;
      front.prev = null;
      if ( !this.offEnd() ) {
        if ( curr.prev == null && curr.next == null ) { curr = null; }
      }
    }
    else { front = back = curr = null; }
    --length;
  }
  
  /* deleteLast()
   * Deletes last element.
   * Pre: !isEmpty().
   */
  void deleteLast() {
    if( this.isEmpty() ){
      throw new RuntimeException("List Error: deleteLast() called on empty List");
    }
    if ( this.length > 1 ) {
      back = back.prev;
      back.next = null;
      if ( !this.offEnd() ) {
        if ( curr.prev == null && curr.next == null ) { curr = null; }
      }
    }
    else { front = back = curr = null; }
    --length;
  }
  
  /* deleteCurrent()
   * Deletes current element.
   * Pre: !isEmpty(), !offEnd(); Post: offEnd().
   */
  void deleteCurrent() {
    if( this.isEmpty() ) {
      throw new RuntimeException("List Error: deleteCurrent() called on empty List");
    }else if ( this.offEnd() ){
      throw new RuntimeException("List Error: deleteCurrent() called on a Null Pointer");
    }
    if ( this.length > 1 ) {
      if ( curr.prev == null ) {
        front = curr.next;
        curr.next = null;
      }else if ( curr.next == null ) {
        back = curr.prev;
        curr.prev = null;
      }else {
        curr.prev.next = curr.next;
        curr.next.prev = curr.prev;
        curr.prev = curr.next = null;
      }
      curr = null;
    }
    else { front = back = curr = null; }
    --length;
  }
  
  // Other methods //////////////////////////////////////////////
  
  /* toString()
   * Overrides Object's toString method and returns a string
   * representation of this List consisting of a space
   * separated list of integers.
   */
  public String toString() {
    String str = "";
    for (Node N = front; N != null; N = N.next){
      str += N.toString() + " ";
    }
    return str;
  }
  
}